### ISSUE TITLE
Optimistic Locking

### ISSUE THREAD

**Issue by Fs02** [Fs02/rel#100]

**Comment by dranikpg** [comment]
There are at least two common approaches to optimistic locking I know:
* using a `where version = ?` clause 
* using a `returning version` and checking versions in code (this works only in transactions)

The first approach is more flexible (can be used in any context), however it doesn't allow us to differentiate between errors for stale and nonexistent records. Is this that big of an issue? I we already have a record with a version, we know it existed (at least in the past). The user would have to get more info about the confict manually on abortion (like querying record with version again)

What about turning locking off? Should it require special functions for saving/deleting? Given, that embedded structs [can be supported](https://github.com/go-rel/rel/pull/262), enabling/disabling could be done with wrapper types.

**Comment by Fs02** [comment]
> The first approach is more flexible (can be used in any context), however it doesn't allow us to differentiate between errors for stale and nonexistent records. Is this that big of an issue? If we already have a record with a version, we know it existed (at least in the past). The user would have to get more info about the confict manually on abortion (like querying record with version again)

yes I agree, the size of issue is depends, but I think it's not possible to differentiate `stale and nonexistent records`, I think we just need to state this explicitly in the documentation

> What about turning locking off?

if user want to update/delete while ignoring lock, user should be able to use `rel.Unscoped()`
example: `repo.Update(ctx, &book, rel.Unscoped())` / `repo.Delete(ctx, &book, rel.Unscoped())`

**Comment by dranikpg** [comment]
Having `repo.Delete(ctx, &model, rel.Unscoped())` is not possible with the current signature, which accepts only a bool vararg for cascades. It it possible to change Delete() to accept mutates or must strong backward compatibility be maintained?

**Comment by Fs02** [comment]
> It it possible to change Delete() to accept mutates or must strong backward compatibility be maintained?

It's okay, this shouldn't break the compilation, and REL is still in unstable version (0.x.x) is due to this possible API change
